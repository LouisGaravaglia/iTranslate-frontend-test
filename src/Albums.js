import React from 'react';
import {Spring} from 'react-spring/renderprops';
import './sass/App.scss';
//COMPONENT IMPORTS.
import SearchResultList from "./SearchResultList";
//REDUX IMPORTS
import {useDispatch, useSelector} from "react-redux";
import {getTracks} from "./actionCreators/BrowseRoute/Artists/getTracksCreator";
import {resetStore} from "./actionCreators/resetStoreCreator";
//CUSTOM HOOK IMPORTS
import useViewport from "./hooks/useViewport";

function Albums({typeOfAlbums}) {
  //REDUX STORE
  const dispatch = useDispatch();
  const albums = useSelector(store => store.albums);
  //VIEWPORT SIZE BASED VARIABLES
  const {viewportWidth} = useViewport();
  let itemsPerPage;
  let albumKey;

  //VIEWPORT BREAKPOINTS TO DETERMINT HOW MANY ALBUM COVERS TO DISPLAY AT ONCE
  if (viewportWidth < 1180 && viewportWidth > 780) {
    itemsPerPage = 2;
  } else if (viewportWidth < 780) {
    itemsPerPage = 1;
  } else {
    itemsPerPage = 3;
  };

  //ALBUM KEY WHICH WOULD CHANGE IF ITEMS PER PAGE CHANGED TO RERENDER COMPONENT IF NUMBER OF ALUMBS DISPLAYING IS CHANGING
  if (itemsPerPage < albums.length) {
    albumKey = itemsPerPage;
  } else {
    albumKey = 0;
  };

////////////////////////////////////////////////////  HANDLE CLICK FUNCTIONS  ////////////////////////////////////////////////////

  const handleAlbumClick = async (albumId) => {
    dispatch(getTracks(albumId));
    dispatch(resetStore("lyrics", "translation"));
  };

////////////////////////////////////////////////////  RETURN  ////////////////////////////////////////////////////

  return (
    <Spring
      from={{opacity: 0}}
      to={{opacity: 1}}
      config={{delay: 300, duration: 300}}
    >
      {props => (
        <div style={props}>

          <div className="Main-Container">
            <SearchResultList key={albums[albumKey].albumId} typeOfResults="albums" resultsArray={albums} handleSearch={handleAlbumClick} itemsPerPage={itemsPerPage} typeOfAlbums={typeOfAlbums}/>
          </div>

        </div>
      )}
    </Spring>
  );
};

export default Albums;