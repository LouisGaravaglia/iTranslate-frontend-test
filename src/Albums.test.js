import React from 'react'
import {render, cleanup} from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import {Provider} from 'react-redux'
import {createStore} from "redux";
import Albums from "./Albums";

afterEach(cleanup);

const startingState = {albums: [{albumId: 42}]};

/**
 * reducer function to mimic the reducer used in Redux.
 * @param {object} state - holds an object of data used within the components
 * @param {object} action - object where the values are strings used to fire certain actions
 */
function reducer(state = startingState, action) {
  switch (action.type) {
    case "ADD_ALBUMs":
      return {...state, albums: action.albums};
    default:
      return state;
  };
};

/**
 * Function that renders the component within a redux environment since most 
 * components rely on redux for data in order to mount.
 * @param {ReactComponent} component - private spotify client id
 * @param {function} store - creates the react store using the createStore method imported from redux
 */
function renderWithRedux(
  component,
  {initialState, store = createStore(reducer, initialState)} = {}
) {
  return {
    ...render(<Provider store={store}>{component}</Provider>)
  };
};

//SMOKE TEST
describe('Smoke Test for Albums component', () => {

  it('renders without crashing', () => {
    renderWithRedux(<Albums />);
  });
});

//SNAPSHOT TEST
describe('Snapshot Test for Albums component', () => {

  it('matches snapshot', () => {
    const {asFragment} = renderWithRedux(<Albums />);
    expect(asFragment()).toMatchSnapshot();
  });
});