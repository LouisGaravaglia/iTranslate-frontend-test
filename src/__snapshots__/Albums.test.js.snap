// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Test for Albums component matches snapshot 1`] = `
<DocumentFragment>
  <div
    style="opacity: 0;"
  >
    <div
      class="Main-Container"
    >
      <div
        class="Browse-Albums"
      >
        <div
          class="Album"
        >
          <span
            class="Hover-Box"
            style="display: inline-block; transform: scale(1);"
          >
            <img
              alt=""
            />
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
